adapter/client/callback.go

```diff
@@ -0,0 +1,80 @@
+package client
+
+import (
+	"path"
+	"regexp"
+	"strings"
+)
+
+var regexpURL = regexp.MustCompile(`(https?)://([^/]+)(/?.*)`)
+
+func callbackDefinitionForClient(c Client) string {
+	callbackDef := ""
+
+	// also support empty clients (for tests without a valid client, using default callbackEndpoint)
+	if c != nil {
+		callbackDef = c.Callback()
+	}
+
+	if callbackDef == "" {
+		// if the callback is empty for the client, use the original behavior:
+		// /oidc/callback is simply appended to the original path, resulting in many different callback URLs
+		callbackDef = "oidc/callback"
+	}
+	return callbackDef
+}
+
```

Comment:
Nit: Why not put the guard first?

```golang
if c == nil {
    // if the callback is empty for the client, use the original behavior:
   // /oidc/callback is simply appended to the original path, resulting in many different callback URLs
   return "oidc/callback"
}

return c.Callback()
```

Reply:
Because I also want to handle the case where there is a non-nil client which doesn't have `callback` attribute set (it is an empty string). In this case, I want it to use the original behavior of relative `oidc/callback`. That way existing users can simply upgrade without braking anything or having to explicitly define the callback.

The nil check has to be there because some routes may not have client defined. It is also safer generally..